src/System.Security.Cryptography.Pkcs/src/Internal/Cryptography/Helpers.cs

```diff
@@ -89,34 +89,82 @@ public static X509Certificate2 TryFindMatchingCertificate(this X509Certificate2C
                 case SubjectIdentifierType.IssuerAndSerialNumber:
                     {
                         X509IssuerSerial issuerSerial = (X509IssuerSerial)(recipientIdentifier.Value);
-
-                        using (ClonedCertificates matches1 = certs.FindCertificates(X509FindType.FindBySerialNumber, issuerSerial.SerialNumber))
+                        string serialNumber = issuerSerial.SerialNumber;
+                        string issuer = issuerSerial.IssuerName;
+                        foreach (X509Certificate2 candidate in certs)
                         {
-                            using (ClonedCertificates matches2 = matches1.FindCertificates(X509FindType.FindByIssuerDistinguishedName, issuerSerial.IssuerName))
-                            {
-                                // Desktop compat: We do not complain about multiple matches. Just take the first one and ignore the rest.
-                                return matches2.Any ? matches2.ClaimResult() : null;
-                            }
+                            if (candidate.SerialNumber == serialNumber && candidate.Issuer == issuer)
```

Comment:
Sounds good.  I'm sure you'll run across it in making the tests work, but watch out for endianness.  (X509Certificate.GetSerialNumber() is LE, I think, but the string is BE).  Based on the fact that X509Certificate2.get_SerialNumber reverses GetSerialNumber().